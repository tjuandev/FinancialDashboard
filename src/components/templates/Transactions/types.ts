import { ComponentProps, ReactNode } from 'react'
import { Header } from 'components/organisms'

export type Props = {
  headerProps: ComponentProps<typeof Header>
  children?: ReactNode
}
