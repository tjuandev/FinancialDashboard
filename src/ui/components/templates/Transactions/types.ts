import { ComponentProps } from 'react'
import { Header } from 'ui/components/organisms'
import { Table } from 'ui/components/molecules'

export type TableProps<TableType> = ComponentProps<typeof Table<TableType>>

export type Props<TableType> = {
  headerProps: ComponentProps<typeof Header>
  tableProps: TableProps<TableType>
}
