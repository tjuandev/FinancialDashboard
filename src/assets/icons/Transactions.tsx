type Props = {
  color?: string
}

function Icon({ color }: Props) {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="20"
      height="26"
      fill="none"
      viewBox="0 0 20 26"
    >
      <path
        stroke={color ? color : 'currentColor'}
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="3"
        d="M18 24V10.25M10 24V2M2 24v-8.25"
      ></path>
    </svg>
  )
}

export default Icon
