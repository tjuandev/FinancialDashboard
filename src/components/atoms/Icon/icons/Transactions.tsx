import { IconProps } from '../types'

function Icon({ color, height = 26, width = 20 }: IconProps) {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width={width}
      height={height}
      fill="none"
      role="img"
      aria-label="Transactions"
      viewBox="0 0 20 26"
    >
      <path
        stroke={color ? color : 'currentColor'}
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="3"
        d="M18 24V10.25M10 24V2M2 24v-8.25"
      ></path>
    </svg>
  )
}

export default Icon
