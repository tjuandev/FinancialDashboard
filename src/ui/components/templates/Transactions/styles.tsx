import { Button } from 'ui/components/atoms'
import styled from 'styled-components'
import theme from 'ui/theme'

export const Container = styled.div`
  padding: 0 ${theme.spacing[20]};
`

export const TransactionsSelectorWrapper = styled.div`
  width: 148px;
`

export const ToolbarContainer = styled.div`
  display: flex;
  justify-content: space-between;

  margin-top: 64px;
  margin-bottom: 24px;

  width: 1143px;
`

export const ToolbarLeftSide = styled.div`
  display: flex;
  gap: 1rem;
`

export const ColumnActionsContainer = styled.div`
  display: flex;
  justify-content: end;
  gap: 0.25rem;

  margin: 0.25rem 0;
`

export const NewEntryButton = styled(Button)`
  width: fit-content;
  display: flex;
  align-items: center;
  gap: 10px;

  padding: 12px 16px;

  :hover {
    filter: brightness(0.9);
  }
`

export const Main = styled.main``
