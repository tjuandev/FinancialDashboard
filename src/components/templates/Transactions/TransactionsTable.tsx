import { createColumnHelper } from '@tanstack/react-table'
import { Button, Icon } from 'components/atoms'
import { Table } from 'components/molecules'
import { css } from 'styled-components'
import theme from 'theme'

import { ColumnActionsContainer } from './styles'

type TransactionsTable = {
  type: string
  date: string
  description: {
    value: string
    columnWidth: string
  }
  value: string
}

const columnHelper = createColumnHelper<TransactionsTable>()

const actionButtonExtraStyles = css`
  width: 40px;
  height: 40px;

  border-radius: ${theme.borderRadius.base};

  :hover {
    filter: brightness(0.9);
  }
`

const tableColumns = [
  columnHelper.accessor('type', {
    header: 'Type',
    cell: (info) => info.getValue()
  }),
  columnHelper.accessor('date', {
    header: 'Date',
    cell: (info) => info.getValue()
  }),
  columnHelper.accessor('description', {
    header: 'Description',
    cell: (info) => info.getValue().value
  }),
  columnHelper.accessor('value', {
    header: 'Value',
    cell: (info) => info.getValue()
  }),
  columnHelper.display({
    id: 'actions',
    header: 'Actions',
    cell: () => (
      <ColumnActionsContainer>
        <Button padding="0.25rem" extraStyles={actionButtonExtraStyles}>
          <Icon name="outlined-trash" />
        </Button>
        <Button padding="0.25rem" extraStyles={actionButtonExtraStyles}>
          <Icon name="outlined-edit" />
        </Button>
      </ColumnActionsContainer>
    )
  })
]

const rows: TransactionsTable[] = [
  {
    type: 'Despesa',
    date: '2022-02-2004',
    description: {
      value: 'Compras no submundo pós apocalíptico de Gânador',
      columnWidth: '100px'
    },
    value: '2000'
  },
  {
    type: 'Despesa',
    date: '2022-02-2004',
    description: {
      value: 'Compras no submundo pós apocalíptico de Gânador',
      columnWidth: '100px'
    },
    value: '2000'
  },
  {
    type: 'Despesa',
    date: '2022-02-2004',
    description: {
      value: 'Compras no submundo pós apocalíptico de Gânador',
      columnWidth: '100px'
    },
    value: '2000'
  }
]

const TransactionsTable = () => {
  return (
    <Table<TransactionsTable>
      columns={tableColumns}
      rows={rows}
      columnsExtensions={[
        {
          id: 'actions',
          horizontalAlign: 'end'
        }
      ]}
    />
  )
}

export default TransactionsTable
