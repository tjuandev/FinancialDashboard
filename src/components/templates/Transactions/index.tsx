import * as S from './styles'

import { Grid, Icon } from 'components/atoms'
import { ExpandableInput, Kpi, Select, Table } from 'components/molecules'
import { Header } from 'components/organisms'

import { Props, TableProps } from './types'

import { IconLiterals } from 'types/icons'

import { v4 as uuid } from 'uuid'

const kpisMock = [
  {
    name: 'R$ 1290,20',
    description: 'Seu saldo atual',
    iconName: 'wallet'
  },
  {
    name: 'R$ 2080,63',
    description: 'Seu ganhos',
    iconName: 'stock-up'
  },
  {
    name: 'R$ 790,43',
    description: 'Suas despesas',
    iconName: 'stock-down'
  }
]

const KPIs = () => (
  <Grid columns="repeat(3, 290px)" gap={8}>
    {kpisMock.map(({ iconName, ...kpiProps }) => (
      <Kpi
        headerChildren={<Icon name={iconName as IconLiterals} />}
        key={uuid()}
        {...kpiProps}
      />
    ))}
  </Grid>
)

const TransactionsSelector = () => (
  <S.TransactionsSelectorWrapper>
    <Select
      defaultValue="all"
      options={[
        { value: 'all', label: 'Todas' },
        { value: 'incomes', label: 'Ganhos' },
        { value: 'outcomes', label: 'Despesas' }
      ]}
    />
  </S.TransactionsSelectorWrapper>
)

const Toolbar = () => (
  <S.ToolbarContainer>
    <TransactionsSelector />
    <S.ToolbarLeftSide>
      <ExpandableInput placeholder="Type something..." />
      <S.NewEntryButton colorSchema="secondary">
        Nova Entrada <Icon name="plus" />
      </S.NewEntryButton>
    </S.ToolbarLeftSide>
  </S.ToolbarContainer>
)

const TransactionsTable = <TableType,>(props: TableProps<TableType>) => {
  const { columns, rows } = props
  const shouldLoad = !columns && !rows

  return <Table loading={shouldLoad} {...props} />
}

const Transactions = <TableType,>({
  headerProps,
  tableProps
}: Props<TableType>) => (
  <S.Container>
    <Header {...headerProps} />
    <S.Main>
      <KPIs />
      <Toolbar />
      <TransactionsTable {...tableProps} />
    </S.Main>
  </S.Container>
)

export default Transactions
