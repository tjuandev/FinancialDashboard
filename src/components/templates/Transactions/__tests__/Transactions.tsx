import { render } from '@testing-library/react'
import Transactions from '..'

describe('<Transactions />', () => {
  it('Should match snapshot', () => {
    const { container } = render(
      <Transactions
        headerProps={{
          avatarProps: {
            name: 'User Test'
          },
          breadcrumbProps: {
            name: 'Test'
          }
        }}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
