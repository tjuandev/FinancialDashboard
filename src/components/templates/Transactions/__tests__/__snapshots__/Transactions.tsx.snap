// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Transactions /> Should match snapshot 1`] = `
.c0 {
  padding: 0 5rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding-top: 1.5rem;
  margin-bottom: 4rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 2rem;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.5rem;
}

.c8 {
  text-align: right;
}

.c8 p {
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 12px;
}

.c8 small {
  font-size: 0.75rem;
}

.c9 {
  width: 48px;
  height: 48px;
  position: relative;
}

.c9,
.c9 img {
  border-radius: 0.625rem;
}

.c10 {
  width: 100%;
  height: 100%;
  border-radius: inherit;
  background: #CDDBC2;
  font-weight: bolder;
  font-size: 1.25rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c6 {
  height: 44px;
  width: 44px;
  background-color: #CDDBC2;
  color: #14404D;
  padding: 0.5rem;
  border-radius: 0.625rem;
  font-size: 1rem;
  font-weight: 600;
  -webkit-transition: all 0.25s ease;
  transition: all 0.25s ease;
  cursor: pointer;
}

.c6:focus {
  outline: 1px solid #14404D;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.5rem;
}

.c3 > h1 {
  font-size: 2rem;
}

.c4 {
  width: 24px;
  height: 24px;
}

<div>
  <div
    class="c0"
  >
    <header
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          />
          <h1>
            Test
          </h1>
        </div>
      </div>
      <div
        class="c5"
      >
        <button
          class="c6"
        >
          <svg
            aria-label="Notification"
            fill="none"
            height="22"
            role="img"
            viewBox="0 0 20 22"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 7A6 6 0 104 7c0 7-3 9-3 9h18s-3-2-3-9zM11.73 20a1.999 1.999 0 01-3.46 0"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <div
          class="c7"
        >
          <div
            class="c8"
          >
            <p>
              User Test
            </p>
            <small />
          </div>
          <div
            class="c9"
          >
            <div
              class="c10"
            >
              UT
            </div>
          </div>
        </div>
      </div>
    </header>
    <main
      class=""
    />
  </div>
</div>
`
